lib/src/rules/always_require_non_null_named_parameters.dart

```diff
@@ -68,50 +68,64 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   @override
   void visitFormalParameterList(FormalParameterList node) {
-    final params = node.parameters
-        // only named parameters
-        .where((p) => p.isNamed)
-        .map((p) => p as DefaultFormalParameter)
-        // without default value
-        .where((p) => p.defaultValue == null)
-        // without @required
-        .where((p) => !p.declaredElement.hasRequired)
-        .toList();
+    List<DefaultFormalParameter> getParams() {
+      final params = <DefaultFormalParameter>[];
+      for (final p in node.parameters) {
+        // Only named parameters
+        if (p.isNamed) {
+          DefaultFormalParameter parameter = p as DefaultFormalParameter;
+          // Without a default value or marked @required
+          if (parameter.defaultValue == null &&
+              !parameter.declaredElement.hasRequired) {
+            params.add(parameter);
+          }
+        }
+      }
+      return params;
+    }
+
     final parent = node.parent;
     if (parent is FunctionExpression) {
-      _checkParams(params, parent.body);
+      _checkParams(getParams(), parent.body);
     } else if (parent is ConstructorDeclaration) {
-      _checkInitializerList(params, parent.initializers);
-      _checkParams(params, parent.body);
+      _checkInitializerList(getParams(), parent.initializers);
+      _checkParams(getParams(), parent.body);
     } else if (parent is MethodDeclaration) {
-      _checkParams(params, parent.body);
+      _checkParams(getParams(), parent.body);
+    }
+  }
+
+  void _checkAsserts(
+      List<Expression> asserts, List<DefaultFormalParameter> params) {
+    for (final expression in asserts) {
+      for (final param in params) {
+        if (_hasAssertNotNull(expression, param.identifier.name)) {
+          rule.reportLintForToken(param.identifier.beginToken);
```

Comment:
Consider removing the parameter from the list after creating a lint. Otherwise redundant asserts will generate multiple diagnostics. It's a fairly uncommon case, so it might not be worth the work from that perspective, but it also has the performance benefit that the list gets shorter making future iterations over the list faster.

Reply:
Interesting idea!

Let's discuss in person.  Thanks!

Reply:
Updated.  Thanks!